go/vt/vtgate/planbuilder/operators/phases.go

```diff
@@ -341,3 +297,93 @@ func addLiteralGroupingToRHS(in *ApplyJoin) (Operator, *ApplyResult) {
 	})
 	return in, NoRewrite
 }
+
+// prepareForAggregationPushing adds columns needed by an operator to its input.
+// This happens only when the filter expression can be retrieved as an offset from the underlying mysql.
+func prepareForAggregationPushing(ctx *plancontext.PlanningContext, root Operator) Operator {
+	addColumnsNeededByFilter := func(in Operator, _ semantics.TableSet, _ bool) (Operator, *ApplyResult) {
+		filter, ok := in.(*Filter)
+		if !ok {
+			return in, NoRewrite
+		}
+
+		var neededAggrs []sqlparser.Expr
+		extractAggrs := func(cursor *sqlparser.CopyOnWriteCursor) {
+			node := cursor.Node()
+			if ctx.IsAggr(node) {
+				neededAggrs = append(neededAggrs, node.(sqlparser.Expr))
+			}
+		}
+
+		for _, expr := range filter.Predicates {
+			_ = sqlparser.CopyOnRewrite(expr, dontEnterSubqueries, extractAggrs, nil)
+		}
+
+		if neededAggrs == nil {
+			return in, NoRewrite
+		}
+
+		addedCols := false
+		aggregator := findAggregatorInSource(filter.Source)
+		for _, aggr := range neededAggrs {
+			if aggregator.FindCol(ctx, aggr, false) == -1 {
+				aggregator.addColumnWithoutPushing(ctx, aeWrap(aggr), false)
+				addedCols = true
+			}
+		}
+
+		if addedCols {
+			return in, Rewrote("added columns because filter needs it")
+		}
+		return in, NoRewrite
+	}
+
+	step1 := TopDown(root, TableID, addColumnsNeededByFilter, stopAtRoute)
+	pushDownHorizon := func(in Operator, _ semantics.TableSet, _ bool) (Operator, *ApplyResult) {
+		// These recursive CTEs have not been pushed under a route, so we will have to evaluate it one the vtgate
+		// That means that we need to turn anything that is coming from the recursion into arguments
+		rcte, ok := in.(*RecurseCTE)
+		if !ok {
+			return in, NoRewrite
+		}
+		hz := rcte.Horizon
+		hz.Source = rcte.Term
+		newTerm, _ := expandHorizon(ctx, hz)
+		pr := findProjection(newTerm)
+		ap, err := pr.GetAliasedProjections()
+		if err != nil {
+			panic(vterrors.VT09015())
+		}
+
+		// We need to break the expressions into LHS and RHS, and store them in the CTE for later use
+		projections := slice.Map(ap, func(p *ProjExpr) *plancontext.RecurseExpression {
+			recurseExpression := breakCTEExpressionInLhsAndRhs(ctx, p.EvalExpr, rcte.LeftID)
+			p.EvalExpr = recurseExpression.RightExpr
+			return recurseExpression
+		})
+		rcte.Projections = projections
+		rcte.Term = newTerm
+		return rcte, Rewrote("expanded horizon on term side of recursive CTE")
+	}
+
+	return TopDown(step1, TableID, pushDownHorizon, stopAtRoute)
+}
+
+func findProjection(op Operator) *Projection {
+	for {
+		proj, ok := op.(*Projection)
+		if ok {
+			return proj
+		}
+		inputs := op.Inputs()
+		if len(inputs) != 1 {
+			panic(vterrors.VT13001("unexpected multiple inputs"))
+		}
+		src := inputs[0]
+		_, isRoute := src.(*Route)
+		if isRoute {
+			panic(vterrors.VT13001("failed to find the projection"))
+		}
+		op = src
+	}
+}
```

Comment:
this looks something like calling 
```
TopDown(root, TableID, visitor, stopAtRoute)
```

probably can use the same method here as well

Reply:
I consciously didn't want to use TopDown, since it can walk down joins, and I don't want that here